### ISSUE TITLE
Port forwarding dos not work (it was dns)

### ISSUE THREAD

**Issue by icepic1984** [k3s-io/k3s#3186]
**Environmental Info:**
K3s Version: 
```
k3s version v1.20.5+k3s1 (355fff30)
go version go1.15.10
```

Node(s) CPU architecture, OS, and Version: 

`Linux kube-node1 5.11.4-1-ARCH #1 SMP Sun Mar 7 23:46:10 UTC 2021 aarch64 GNU/Linux`

on a Raspberry Pi 4.

Cluster Configuration:
1 Master and one node

**Describe the bug:**
Port forwarding to test pod network services does not work. Steps to reproduce the problem:

1)  Execute  ` kubectl create deployment nginx --image=nginx`
2)  `kubectl port-forward deployment/nginx 8080:80`
3) `curl localhost:8080`

Output of `kubectl port-forward`:
```
Handling connection for 8080
E0413 12:58:10.516183  263245 portforward.go:400] an error occurred forwarding 8080 -> 80: error forwarding port 80 to pod 7ce8494ab57cab112516fd11aefde5f8aed0b66f7ce2935f0e3a3260ccc0a62e, uid : failed to execute portforward in network namespace "/var/run/netns/cni-e94be3ff-2e9e-d42f-c1c2-9220ebd5a2b6": failed to dial 80: dial tcp4: lookup localhost: Try again
```

`curl` will time out. From the output, it seems like `localhost` cannot be resolved.

However, when i create a `ClusterIp` type service and log into a busybox pod, i can reach the network service without problems. Exposing the service via `NodePort` type also works flawless. There are no suspicious log entries in `k3s.service` or `k3s-agent.service`

**Additional context / logs:**
```
$sudo k3s check-config

Verifying binaries in /var/lib/rancher/k3s/data/e8aa7a486139535c59667cafb62d3f38849345f5e3f61128cf2a63a3c848bc97/bin:
- sha256sum: good
- links: good

System:
- /usr/bin iptables v1.8.7 (legacy): ok
- swap: disabled
- routes: default CIDRs 10.42.0.0/16 or 10.43.0.0/16 already routed

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000

info: reading kernel config from /proc/config.gz ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_IPVS: enabled (as module)
- CONFIG_IP_NF_NAT: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_POSIX_MQUEUE: enabled

Optional Features:
- CONFIG_USER_NS: enabled
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: enabled
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: enabled
- CONFIG_NET_CLS_CGROUP: enabled
- CONFIG_CGROUP_NET_PRIO: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: missing
- CONFIG_IP_NF_TARGET_REDIRECT: enabled (as module)
- CONFIG_IP_SET: enabled (as module)
- CONFIG_IP_VS: enabled (as module)
- CONFIG_IP_VS_NFCT: enabled
- CONFIG_IP_VS_PROTO_TCP: enabled
- CONFIG_IP_VS_PROTO_UDP: enabled
- CONFIG_IP_VS_RR: enabled (as module)
- CONFIG_EXT4_FS: enabled
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
      Optional (for encrypted networks):
      - CONFIG_CRYPTO: enabled
      - CONFIG_CRYPTO_AEAD: enabled
      - CONFIG_CRYPTO_GCM: enabled
      - CONFIG_CRYPTO_SEQIV: enabled
      - CONFIG_CRYPTO_GHASH: enabled
      - CONFIG_XFRM: enabled
      - CONFIG_XFRM_USER: enabled
      - CONFIG_XFRM_ALGO: enabled
      - CONFIG_INET_ESP: enabled (as module)
      - CONFIG_INET_XFRM_MODE_TRANSPORT: missing
- Storage Drivers:
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)

```
```
$kubectl version

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5+k3s1", GitCommit:"355fff3017b06cde44dbd879408a3a6826fa7125", GitTreeState:"clean", BuildDate:"2021-03-31T06:20:03Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/arm64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4+k3s1", GitCommit:"838a906ab5eba62ff529d6a3a746384eba810758", GitTreeState:"clean", BuildDate:"2021-02-22T19:49:35Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/arm64"}

```

Any tips how to debug this issues are welcome.

**Comment by dweomer** [comment]
Because k3s (and other k8s bits) are go programs and not yet compiled with go 1.16 that purports to resolve this issue (pun alert!) you will need a minimal `/etc/nsswitch.conf` on your system (or in the container k3s happens to be running in). e.g.:
```bash
# docker pull docker:latest
$ docker run --rm -it docker cat /etc/nsswitch.conf
hosts: files dns
```

See also https://github.com/rancher/k3os/blob/v0.20.4-k3s1r0/overlay/etc/nsswitch.conf

**Comment by dweomer** [comment]
![image](https://user-images.githubusercontent.com/1033444/114589972-0bcf3880-9c3d-11eb-82ca-df0c7f02bdfb.png)

**Comment by icepic1984** [comment]
Hm this really sounds like it could be my problem, especially after reading [this](https://medium.com/box-tech-blog/a-trip-down-the-dns-rabbit-hole-understanding-the-role-of-kubernetes-golang-libc-systemd-41fd80ffd679). However, i wasn't able to fix my problem. What i  did:

1) Change `/etc/nsswitch.conf` on master and node (see below). I run k3s baremetal not in a container.
2) Reboot both systems, because i wasn't sure, if restarting the serivces is enough.

I changed the /etc/nsswitch.conf to

```
# Name Service Switch configuration file.
# See nsswitch.conf(5) for details.

passwd: files systemd
group: files [SUCCESS=merge] systemd
shadow: files

publickey: files

#hosts: files mymachines myhostname dns resolve [!UNAVAIL=return]
hosts: files dns

networks: files

protocols: files
services: files
ethers: files
rpc: files

netgroup: files

```
See the host line (the one commented out was the original entry). From my understanding the rest of the entries are irrelevant to my problem.

However, unfortunately the exact same problem still occurs. Running `getent` on master or node results in:

```
 getent ahostsv4 localhost
127.0.0.1       STREAM localhost
127.0.0.1       DGRAM
127.0.0.1       RAW
127.0.0.1       STREAM
127.0.0.1       DGRAM
127.0.0.1       RAW
```

which looks correct to me. Additionally i tried adding a  `/etc/hosts` file, which was not provided by my dist. F

```
# Static table lookup for hostnames.
# See hosts(5) for details.
127.0.0.1 localhost
::1 localhost
```
No luck either. Any more ideas?

**Comment by alexandresoro** [comment]
Hello,

I stumbled on this issue while having a similar issue with port forwarding, on the same environment (ArchLinux host).

After hours of investigation, I finally managed to solve it. In my case, my `/etc/hosts` was empty. Adding entries for localhost as mentioned above solved it for me.

Apologies is this is not relevant, but I felt that the issue was similar to mine, but it may be helpful to know that setting up the localhost entries seem to be required.
